Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm

```diff
@@ -254,7 +254,15 @@
 bool WebExtensionTab::extensionHasPermission() const
 {
     ASSERT(extensionHasAccess());
-    return extensionContext()->hasPermission(url(), const_cast<WebExtensionTab*>(this));
+
+    auto tabURL = url();
+
+    if (!tabURL.protocolIsInHTTPFamily()) {
+        // For non-HTTP family URLs, extensions should only have access to tabs with their own extension's base URL.
+        return extensionContext()->isURLForThisExtension(tabURL);
+    }
+
+    return extensionContext()->hasPermission(tabURL, const_cast<WebExtensionTab*>(this));
```

Comment:
You can revert this, since `extensionContext()->hasPermission(tabURL, …)` already checks `isURLForThisExtension`.

```c++
    if (isURLForThisExtension(url))
        return PermissionState::GrantedImplicitly;
```

Just adding the check for extensionHasPermission() is enough.

Reply:
And then after that, there is:

```
    if (!WebExtensionMatchPattern::validSchemes().contains(url.protocol().toStringWithoutCopying()))
        return PermissionState::Unknown;
```

So any other `extension-scheme://` URLs will fail that check.

Reply:
Oh, but we add the custom given extension schemes to the `validSchemes`, but I don't think there would be any way to grant them permissions, so I think this should still be safe.

Reply:
Yep, a third-party app could use the WebKit API to grant them explicitly I suppose, But that would likely be best to be blocked at the API level to prevent undefined behavior.